projects/project2/src/main/scala/com/kg/carl/Algorithm.scala

```diff
@@ -0,0 +1,280 @@
+package com.kg.carl
+import com.kg.carl.Utils._
+import util.control.Breaks._
+import scala.collection.mutable._
+import com.kg.carl.Preprocessor._
+
+object Algorithm {
+
+  /**
+   * the case class for scored rules.
+   * all metrics declared will be evaluated/computed with the CARL algorithm
+   * the rule can be viewed as p(x,y) ∧ q(y,z) -> r(x,z)
+   *
+   * @param p - ID for p node entity
+   * @param q - ID for q node entity
+   * @param r - ID for r node entity
+   */
+  case class ScoredRule(
+    p: Int,
+    q: Int,
+    r: Int) {
+    var support = 0
+    var bodySupport = 0
+    var headCoverage = 0.0
+    var standardConfidence = 0.0
+    var pcaConfidence = 0.0
+    var completenessConfidence = 0.0
+    var precision = 0.0
+    var recall = 0.0
+    var directionalMetric = 0.0
+    var directionalCoef = 0.0
+  }
+  /**
+   * declaring metric standards as mentioned in the paper
+   */
+  val MIN_HEAD_COVERAGE = 0.001
+  val MIN_STANDARD_CONFIDENCE = 0.001
+  val MIN_SUPPORT = 10
+  val CONFIDENCE_INCOMPLETENESS_FACTOR = 0.5
+
+  /**
+   * Mining in KGs have high degree of incompleteness, which may provide
+   * inaccurate quality of mined rules, The effort of this algorithm is to
+   * expose that incompleteness by introducing aware scoring functions.
+   * First the algorithm count number of triples per relations and number of
+   * entities, then count the number of missing triples per relation and
+   * compute support and other metrics for each possible rule and body.
+   * Note: This algorithm is not dependent on any 3rd party library such AMIE.
+   * Also, does not use methodology from any other Association Rule Learning.
+   * It is primitive and implements the work mentioned in the paper.
+   *
+   * @param pso	- PredicateSubjectObject map
+   * @param nodes	- list of nodes
+   * @param properties	- a list of all ids
+   * @param ecpv - expected cardinality by property value
+   * @param numRules - max number of rules to be scored
+   * @return  - list of scored rules
+   */
+  def CARLAlgorithm(
```

Comment:
Will try to comment only on Spark implementation of the algorithm. So will leave this open for you to comment during the presentation.